### PR TITLE
[JENKINS-31155] Ability to retrieve a single revision

### PR DESCRIPTION
**Pull request by jglick**
Downstream of https://github.com/jenkinsci/scm-api-plugin/pull/12.

@reviewbybees

### CODE REVIEW

**Comment by jglick** [comment]
Oops, forgot to deploy the parent POM snapshot.

File: pom.xml [hunk 1]
```diff
@@ -173,7 +173,7 @@
     <dependency>
       <groupId>org.jenkins-ci.plugins</groupId>
       <artifactId>scm-api</artifactId>
-      <version>1.0</version>
+      <version>1.3-SNAPSHOT</version> <!-- TODO https://github.com/jenkinsci/scm-api-plugin/pull/12 -->
```
Comment (MarkEWaite) [inline pom.xml]: @stephenc, I assume that this pull request needs to wait for the next release of scm-api.

Comment (jglick) [inline pom.xml]: Yes, but @stephenc is on vacation for another week so I am considering pushing for a https://github.com/jenkinsci/scm-api-plugin/pull/12 release before then.

File: src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java [hunk 2]
```diff
@@ -185,19 +188,31 @@ protected SCMRevision retrieve(@NonNull SCMHead head, @NonNull TaskListener list
...
             client.fetch(remoteName, refSpecs.toArray(new RefSpec[refSpecs.size()]));
-            // we don't prune remotes here, as we just want one head's revision
-            for (Branch b : client.getRemoteBranches()) {
-                String branchName = StringUtils.removeStart(b.getName(), remoteName + "/");
-                if (branchName.equals(head.getName())) {
-                    return new SCMRevisionImpl(head, b.getSHA1String());
-                }
-            }
-            return null;
+            return retriever.run(client, remoteName);
```
Comment (jglick) [inline src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java]: Big diff blocks are just removing duplication present in @stephenc’s original implementation, preparing to use `doRetrieve` some more times.

File: src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java [hunk 3]
```diff
@@ -262,97 +286,113 @@ private void _release(RevWalk walk) {
...
-                    if (branchCriteria != null) {
-                        RevCommit commit = walk.parseCommit(b.getSHA1());
-                        final long lastModified = TimeUnit.SECONDS.toMillis(commit.getCommitTime());
-                        final RevTree tree = commit.getTree();
-                        SCMSourceCriteria.Probe probe = new SCMSourceCriteria.Probe() {
-                            @Override
-                            public String name() {
-                                return branchName;
-                            }
+        doRetrieve(new Retriever<Void>() {
```
Comment (jglick) [inline src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java]: Easier to follow when [ignoring indentation changes](https://github.com/jenkinsci/git-plugin/pull/433/files?w=1#diff-33b2f30e4119d2fd0e0bb8653c0c58bdR291).

**Comment by abayer** [comment]
:bee:

**Comment by abayer** [comment]
re-:bee:

**Comment by jglick** [comment]
@reviewbybees done

**Comment by jglick** [comment]
>  I assume that this pull request needs to wait for the next release of scm-api

Released and dependency updated to match.

**Comment by jglick** [comment]
@MarkEWaite is this & #437 still under consideration?

**Comment by stephenc** [comment]
LGTM 👍

**Pull request merged**